### PR TITLE
feat(codeeditor): support to extensions

### LINKED ISSUE #280: Feature request: vim keybindings
# Tell us how you think we can improve Sandpack

It would be cool to see vim keybindigns

## Packages affected

- [x] sandpack-client
- [ ] sandpack-react

## What is this feature?

Vim keybindings

## How would your idea work?

It would allow you to use Vim

## Do you have any examples of how you would like to see us implement it?

Here is how replit does it for Codemirror: https://github.com/replit/codemirror-vim

### PR DESCRIPTION
**Pull request by danilowoz**
Introduces the ability to load Codemirror extensions in the CodeEditor component, in two ways:

- By SandpackCodeEditor props:
```jsx
<SandpackProvider template="react">
  <SandpackThemeProvider>
    <SandpackCodeEditor
      extensions={[autocompletion()]}
      extensionsKeymap={[completionKeymap]}
    />
  </SandpackThemeProvider>
</SandpackProvider>
```

- By Sandpack props:
```jsx
<Sandpack
  options={{
    codeEditor: {
      extensions: [autocompletion()],
      extensionsKeymap: [completionKeymap],
    },
  }}
  template="react"
/>
```

Closes #280

### CODE REVIEW

**Review by jsjoeio — APPROVED** [review]
Looks great! Thanks for adding this so quickly 😄

A couple nits around capitalization of CodeMirror but seems like it's good to go

bb8 thumbs up

File: sandpack-react/src/components/CodeEditor/index.tsx [hunk 1]
```diff
@@ -27,6 +29,16 @@ export interface CodeEditorProps {
...
+  /**
+   * Codemirror extensions for the editor state, which can
+   * provide extra features and functionalities to the editor component.
+   */
+  extensions?: Extension[];
+  /**
+   * Property to register Codemirror extension keymap.
+   */
+  extensionsKeymap?: Array<readonly KeyBinding[]>;
+  id?: string;
```
[the remark this answers is not in the document]
Comment (danilowoz) [reply sandpack-react/src/components/CodeEditor/index.tsx:41]: Good question. This is a temp fix that I'll approach in another PR soon. It's related to a random id generated in every render, which leads to unmatch values in tests cases. Here's more detail https://github.com/codesandbox/sandpack/issues/143

File: cypress/integration/CodeMirror.spec.js [hunk 2]
```diff
@@ -20,4 +20,18 @@ describe("CodeMirror", () => {
 
     cy.get(".cm-content").invoke("attr", "style", "tab-size: 4;").snapshot();
   });
+
+  it("should load a codemirror extension", () => {
```
[the remark this answers is not in the document]
Comment (danilowoz) [reply cypress/integration/CodeMirror.spec.js:24]: Reading it now, indeed this test is not very clear. Let me improve it and leave more details what is suppose to test.

File: website/docs/docs/advanced-usage/components.md [hunk 3]
```diff
@@ -153,6 +153,31 @@ If you played with the `Sandpack` preset, you should be familiar already with th
   </SandpackLayout>
 </SandpackProvider>
 
+### Extensions
+
+Sandpack uses CodeMirror behind the hood to provide a nice editor. You can extend the editor with your own extensions, like [`@codemirror/autocomplete`](https://www.npmjs.com/package/@codemirror/autocomplete).
```
Comment (sannek) [inline website/docs/docs/advanced-usage/components.md:158]: `behind the scenes` or `under the hood` :) Also, I guess there isn't a nice list of codemirror extensions anywhere? I looked for it, but I couldn't find it. If you do know of one, a link to such a list might be nice?

```suggestion
Sandpack uses CodeMirror under the hood to provide a nice editor. You can extend the editor with any CodeMirror extensions, such as [`@codemirror/autocomplete`](https://www.npmjs.com/package/@codemirror/autocomplete).
```

Comment (danilowoz) [reply website/docs/docs/advanced-usage/components.md:158]: Oh, thanks for clarifying it to me! 

Yeah, I also haven't found any interesting extension list, that's why I pointed to an official one that I used in the examples.

**Comment by jsjoeio** [comment]
Any ideas on when this will be merged?

**Pull request merged**